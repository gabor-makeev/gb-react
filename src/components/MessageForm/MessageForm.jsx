import style from './MessageForm.module.scss';
import { AUTHORS, STYLES } from '../../constants';
import { useRef, useState } from 'react';

export const MessageForm = ({
  pushMessage,
  border = STYLES.border,
  borderRadius = STYLES.borderRadius,
}) => {
  const [messageContent, setMessageContent] = useState('');

  const messageFormStyle = {
    border: border,
    borderRadius: borderRadius,
  };

  const input = useRef();

  const sendMessage = (e) => {
    e.preventDefault();
    pushMessage({
      text: messageContent,
      author: AUTHORS.user,
    });

    setMessageContent('');
    input.current.focus();
  };

  return (
    <form
      onSubmit={sendMessage}
      className={style['message-form']}
      style={messageFormStyle}
    >
      <input
        type="text"
        value={messageContent}
        ref={input}
        className={style['message-form__input']}
        onChange={(e) => {
          setMessageContent(e.target.value);
        }}
      />
      <button
        className={style['message-form__button']}
        disabled={!messageContent}
      >
        Send
      </button>
    </form>
  );
};
