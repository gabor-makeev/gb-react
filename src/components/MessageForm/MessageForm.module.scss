.message-form {
  background-color: #FF4F4F;
  width: 300px;
  box-shadow: rgba(0, 0, 0, 0.35) 0 5px 15px;
  padding: 20px;
  display: flex;
  gap: 10px;
  justify-content: space-between;
}