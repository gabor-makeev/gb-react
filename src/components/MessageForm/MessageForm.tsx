import style from './MessageForm.module.scss';
import { AUTHORS, STYLES } from 'src/constants';
import { MessageItem } from 'src/default-types';
import React, { FC, useState } from 'react';

import { Button } from './Button/Button';
import { Input } from './Input/Input';
import { Container } from '@mui/material';

interface MessageFormProps {
  setMessageList: (newMessageList: MessageItem[]) => void;
  messageList: MessageItem[];
  border?: string;
  borderRadius?: string;
}

export const MessageForm: FC<MessageFormProps> = ({
  setMessageList,
  messageList,
  border = STYLES.border,
  borderRadius = STYLES.borderRadius,
}) => {
  const [inputValue, setInputValue] = useState('');

  const formContainerStyle = {
    border: border,
    borderRadius: borderRadius,
    backgroundColor: '#CCCCCC',
    width: '300px',
    boxShadow: 'rgba(0, 0, 0, 0.35) 0 5px 15px',
    padding: '20px',
    display: 'flex',
    gap: '10px',
    justifyContent: 'space-between',
    margin: '0',
  };

  const sendMessage = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    setMessageList([
      ...messageList,
      {
        text: inputValue,
        author: AUTHORS.user,
      },
    ]);

    setInputValue('');
  };

  return (
    <form
      onSubmit={sendMessage}
      className={style['message-form']}
      style={{ width: '100%' }}
    >
      <Container sx={formContainerStyle}>
        <Input value={inputValue} setValue={setInputValue} />
        <Button disabled={!inputValue} />
      </Container>
    </form>
  );
};
