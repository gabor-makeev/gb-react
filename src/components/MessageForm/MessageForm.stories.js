import { MessageForm } from './MessageForm';

export default {
  title: 'MessageForm',
  component: MessageForm,
};

const Template = (args) => <MessageForm {...args} />;

export const Primary = Template.bind({});

export const BoldBorder = Template.bind({});
BoldBorder.args = {
  border: '5px solid black',
};

export const SquareBorders = Template.bind({});
SquareBorders.args = {
  borderRadius: 'unset',
};
