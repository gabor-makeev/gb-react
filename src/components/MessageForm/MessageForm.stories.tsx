import { MessageForm } from './MessageForm';
import { ComponentMeta, ComponentStory } from '@storybook/react';

export default {
  title: 'MessageForm',
  component: MessageForm,
} as ComponentMeta<typeof MessageForm>;

const Template: ComponentStory<typeof MessageForm> = (args) => (
  <MessageForm {...args} />
);

export const Primary = Template.bind({});

export const BoldBorder = Template.bind({});
BoldBorder.args = {
  border: '5px solid black',
};

export const SquareBorders = Template.bind({});
SquareBorders.args = {
  borderRadius: 'unset',
};
