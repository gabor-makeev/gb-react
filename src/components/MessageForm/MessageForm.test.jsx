import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { MessageForm } from './MessageForm';
import userEvent from '@testing-library/user-event';

describe('MessageForm', () => {
  it('should render', () => {
    render(<MessageForm />);
  });

  it('should have the button disabled when no input entered', () => {
    render(<MessageForm />);

    expect(screen.getByRole('button')).toBeDisabled();
  });

  it('should have clear input field after message sent', async () => {
    const mockHandler = jest.fn();
    render(<MessageForm pushMessage={mockHandler} />);
    const input = screen.getByDisplayValue('');

    await userEvent.type(input, 'test');
    await userEvent.click(screen.getByRole('button'));

    expect(input.value).toBe('');
  });

  it('should render with snapshot', () => {
    const { asFragment } = render(
      <MessageForm border={'2px solid blue'} borderRadius={'25px'} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
