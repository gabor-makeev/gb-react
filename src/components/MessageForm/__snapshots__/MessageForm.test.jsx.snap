// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageForm should render with snapshot 1`] = `
<DocumentFragment>
  <form
    class="message-form"
    style="border: 2px solid blue; border-radius: 25px;"
  >
    <input
      class="input"
      type="text"
      value=""
    />
    <button
      class="button"
      disabled=""
    >
      Send
    </button>
  </form>
</DocumentFragment>
`;
